File: practice/data_structures/heap/qheap1/qheap1.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>

using namespace std;

class Heap
{
public:
    Heap()
        : m_size(0)
    {}
    
    void insert(int i_val)
    {
        m_data[++m_size] = i_val;
        m_pos[i_val] = m_size;
        if (m_size > 0)
        {
            int pos = m_size;
            while (pos > 1)
            {
                if (m_data[pos] < m_data[pos / 2])
                {
                    m_pos[m_data[pos]] = pos / 2;
                    m_pos[m_data[pos / 2]] = pos;
                    swap(m_data[pos], m_data[pos / 2]);
                    pos /= 2;
                }
                else
                {
                    break;
                }
            }
        }
    }
    
    void remove(int i_val)
    {
        int idx = m_pos[i_val];
        m_pos[i_val] = 0;
        m_pos[m_data[m_size]] = idx;
        m_data[idx] = m_data[m_size--];
        
        while (true)
        {
            int l = 2 * idx, r = 2 * idx + 1;
            if (l <= m_size)
            {
                if (r < m_size)
                {
                    if (m_data[idx] > m_data[l] || m_data[idx] > m_data[r])
                    {
                        int sidx = (m_data[l] < m_data[r]) ? l : r;
                        m_pos[m_data[sidx]] = idx;
                        m_pos[m_data[idx]] = sidx;
                        swap(m_data[idx], m_data[sidx]);
                        idx = sidx;
                    }
                    else
                    {
                        break;
                    }
                }
                else
                {
                    if (m_data[idx] > m_data[l])
                    {
                        m_pos[m_data[l]] = idx;
                        m_pos[m_data[idx]] = l;
                        swap(m_data[idx], m_data[l]);
                        idx = l;
                    }
                    else
                    {
                        break;
                    }
                }
            }
            else
            {
                break;
            }
        }
    }
    
    int min_val() const
    {
        return m_data[1];
    }
    
private:
    map<int, int> m_pos;
    int m_data[100000];
    int m_size;
};

int main()
{
    int q, v, t;
    Heap heap;
    
    cin >> q;
    while (q--)
    {
        cin >> v;
        if (v == 1)
        {
            cin >> t;
            heap.insert(t);
        }
        else if (v == 2)
        {
            cin >> t;
            heap.remove(t);
        }
        else
        {
            cout << heap.min_val() << endl;
        }
    }
    
    return 0;
}

File: practice/data_structures/heap/qheap1/qheap1.py
#!/usr/bin/env python

class Heap(object):
    def __init__(self):
        self.__Pos = dict()
        self.__Data = [0 for _ in range(100000)]
        self.__Size = 0
        
    def insert(self, i_val):
        self.__Size += 1
        self.__Data[self.__Size] = i_val
        self.__Pos[i_val] = self.__Size
        
        pos = self.__Size
        while pos > 1:
            if self.__Data[pos] < self.__Data[pos // 2]:
                self.__Pos[self.__Data[pos]] = pos // 2
                self.__Pos[self.__Data[pos // 2]] = pos
                self.__Data[pos], self.__Data[pos // 2] = self.__Data[pos // 2], self.__Data[pos]
                pos //= 2
            else:
                break
    
    def remove(self, i_val):
        idx = self.__Pos[i_val]
        self.__Pos[i_val] = 0
        self.__Pos[self.__Data[self.__Size]] = idx
        self.__Data[idx] = self.__Data[self.__Size]
        self.__Size -= 1
        
        while True:
            l, r = 2 * idx, 2 * idx + 1
            if l <= self.__Size:
                if r < self.__Size:
                    if self.__Data[idx] > self.__Data[l] or self.__Data[idx] > self.__Data[r]:
                        sidx = l if self.__Data[l] < self.__Data[r] else r
                        self.__Pos[self.__Data[sidx]] = idx
                        self.__Pos[self.__Data[idx]] = sidx
                        self.__Data[idx], self.__Data[sidx] = self.__Data[sidx], self.__Data[idx]
                        idx = sidx
                    else:
                        break
                else:
                    if self.__Data[idx] > self.__Data[l]:
                        self.__Pos[self.__Data[l]] = idx
                        self.__Pos[self.__Data[idx]] = l
                        self.__Data[idx], self.__Data[l] = self.__Data[l], self.__Data[idx]
                        idx = l
                    else:
                        break
            else:
                break
    
    def min_val(self):
        return self.__Data[1]


q = int(raw_input().strip())
heap = Heap()
for _ in xrange(q):
    i = map(int, raw_input().strip().split())
    if i[0] == 1:
        heap.insert(i[1])
    elif i[0] == 2:
        heap.remove(i[1])
    else:
        print heap.min_val()

